app/javascript/utils/lazy-highlight-all.ts

```diff
@@ -0,0 +1,46 @@
+export function lazyHighlightAll(): void {
+  let highlighted = false
+
+  function loadAndHighlight() {
+    import('@/utils/highlight')
+      .then((m) => {
+        m.highlightAll()
+        observer.disconnect()
+      })
+      .catch((e) => {
+        observer.disconnect()
+        throw new Error(e)
+      })
+  }
+
+  function checkAndHighlight(): void {
+    // only load highlightjs when a code block exists in the DOM
+    if (highlighted || !hasHighlightableCodeBlock(document)) return
+
+    // to avoid race conditions, set this optimistically to true before the async loading
+    highlighted = true
+    loadAndHighlight()
+  }
+
+  function mutationCallback(mutationsList: MutationRecord[]): void {
+    for (const mutation of mutationsList) {
+      if (mutation.type !== 'childList') continue
+      for (const node of mutation.addedNodes) {
+        if (node.nodeType !== Node.ELEMENT_NODE) continue
+        if (hasHighlightableCodeBlock(node as Element)) {
+          checkAndHighlight()
+          return
```

Comment:
This bit is a little messy, as it calls `checkAndHighlight` which then does another check for `hasHighlightableCodeBlock` on the whole document.

I'd suggest either:
1. Passing a `force` attribute (set to `false` by default) into checkAndHighlight to avoid this; or
2. Break out L21-22 into a different function that's called here and in checkAndHighlight().

Option 2 is probably better